### DISCUSSION TITLE
Yii2 Debug toolbar - allow DB logs with an admin account

### DISCUSSION THREAD

**Discussion by Anubarak** [craftcms/cms#14985] [Ideas] [answered]
I build this functionality for my Craft 3 projects via events and different config values but it seems Craft 4/5 made it nearly unable to activate logging with an admin account without heavy customizations.

I often need to use the Yii2 Debug toolbar in production but the most important thing - the logs, especially the DB logs - are missing.
I was able to overwrite most things to re-enable logging, however I do not want to overwrite too many classes just because of that.

It would be awesome if there was a native setting for user preferences to include profiles and logs in the debug toolbar while in production mode

**Comment by brandonkelly** [comment, marked as answer]
Since Craft 4, DB profiling and logging are only enabled for Dev Mode by default, but you can customize it by overriding the `db` component config in `config/app.php`:

```php
return [
    'components' => [
        'db' => function() {
            $config = craft\helpers\App::dbConfig();
            $config['enableLogging'] = true;
            $config['enableProfiling'] = true;
            return Craft::createObject($config);
        },
    ],
];
```

**Reply by Anubarak** [reply]
I already tried that and did the best I can otherwise I wouldn't have started this discussion.
Unfortunately most things aren't logged as `Yii2::debug` function includes a `YII_DEBUG` check.

Yii2 uses `Yii::info` to log queries when they are executed
https://github.com/yiisoft/yii2/blob/2.0.49.3/framework/db/Command.php#L1145
```php
    /**
     * Logs the current database query if query logging is enabled and returns
     * the profiling token if profiling is enabled.
     * @param string $category the log category.
     * @return array array of two elements, the first is boolean of whether profiling is enabled or not.
     * The second is the rawSql if it has been created.
     */
    protected function logQuery($category)
    {
        if ($this->db->enableLogging) {
            $rawSql = $this->getRawSql();
            Yii::info($rawSql, $category);
        }
        if (!$this->db->enableProfiling) {
            return [false, isset($rawSql) ? $rawSql : null];
        }

        return [true, isset($rawSql) ? $rawSql : $this->getRawSql()];
    }
```


and the info function directly logs it
https://github.com/yiisoft/yii2/blob/2.0.49.3/framework/BaseYii.php#L457-L460

```PHP
    /**
     * Logs an informative message.
     * An informative message is typically logged by an application to keep record of
     * something important (e.g. an administrator logs in).
     * @param string|array $message the message to be logged. This can be a simple string or a more
     * complex data structure, such as an array.
     * @param string $category the category of the message.
     */
    public static function info($message, $category = 'application')
    {
        static::getLogger()->log($message, Logger::LEVEL_INFO, $category);
    }
```

While Craft uses `Craft::debug`
https://github.com/craftcms/cms/blob/4.9.4/src/db/Command.php#L279-L290
```PHP
    protected function logQuery($category): array
    {
        if ($this->db->enableLogging) {
            $rawSql = $this->getRawSql();
            Craft::debug("SQL query:\n" . $rawSql, $category);
        }
        if (!$this->db->enableProfiling) {
            return [false, isset($rawSql) ? $rawSql : null];
        }

        return [true, isset($rawSql) ? $rawSql : $this->getRawSql()];
    }
```

and these one are aren't logged due to YII_DEBUG check https://github.com/yiisoft/yii2/blob/2.0.49.3/framework/BaseYii.php#L404-L409
```php
    /**
     * Logs a debug message.
     * Trace messages are logged mainly for development purposes to see
     * the execution workflow of some code. This method will only log
     * a message when the application is in debug mode.
     * @param string|array $message the message to be logged. This can be a simple string or a more
     * complex data structure, such as an array.
     * @param string $category the category of the message.
     * @since 2.0.14
     */
    public static function debug($message, $category = 'application')
    {
        if (YII_DEBUG) {
            static::getLogger()->log($message, Logger::LEVEL_TRACE, $category);
        }
    }
```

So basically yes: I can enable it, but the relevant DB Queries aren't logged.

I know I could use DI to use a different `Command` class but there are a few other places as well and I do not want to mess to much with your internal classes.
